net_stats.c

```diff
@@ -0,0 +1,168 @@
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <linux/sockios.h>
+#include <linux/types.h>
+#include <linux/ethtool.h>
+
+#include "system_stats.h"
+#include "net_stats.h"
+
+#ifndef	DUPLEX_UNKNOWN
+#define	DUPLEX_UNKNOWN		0xFF
+#endif /* DUPLEX_UNKNOWN */
+
+extern system_stat system_stats_old;
+
+static net_stats net_stats_old = {0,};
+static net_stats net_stats_next = {0,};
+
+static void read_net_stats(net_stats *ns)
+{
+	char *p;
+	char buf[256], name[IF_NAMESIZE + 1];
+	unsigned long long rx_bytes, rx_packets, rx_errors, tx_bytes, tx_packets, tx_errors, collisions, sat[4];
+	int n, j = 0;
+	net_stat *stats = ns->values;
+	FILE *net = fopen("/proc/net/dev", "r");
+
+	if (net == NULL) return;
+
+	while (fgets(buf, sizeof(buf), net)) {
+		if (j++ < 2) continue; /* skip headers */
+		if (sscanf(buf, "%s %llu %llu %llu %llu %*u %*u %*u %*u %llu %llu %llu %llu %llu %llu %llu %*u",
+					name, &rx_bytes, &rx_packets, &rx_errors, &sat[0], &tx_bytes, &tx_packets,
+					&tx_errors, &sat[1], &sat[2], &collisions, &sat[3]) == 12) {
+			/* Skip interface if it has never seen a packet */
+			if (rx_packets == 0 && tx_packets == 0) continue;
+
+			if ((p = strrchr(name, ':')))
+				*p = '\0';
+
+			for (n = 0; n < ns->size; ++n)
+				if (strcmp(stats[n].name, name) == 0)
+					break;
+
+			if (n == ns->size) { /* interface is not in our list */
+				int sock;
+				if (ns->size++ >= ns->len)
+					stats = ns->values = repalloc(ns->values, sizeof(net_stat)*(ns->len = ns->size));
+
+				memset(stats + n, 0, sizeof(net_stat));
+				strcpy(stats[n].name, name);
```

Comment:
I want to suggest to use strncpy, since we know the string size limit (otherwise there is a room for overruns if somehow there is no null byte).

Reply:
I was assuming that length of network interface name is guaranteed to fit into `IF_NAMESIZE`.